linux/ibus-keyman/src/engine.c

```diff
@@ -576,34 +600,63 @@ ibus_keyman_engine_process_key_event (IBusEngine     *engine,
         switch(action_items[i].type)
         {
             case KM_KBP_IT_CHAR:
-                numbytes = g_unichar_to_utf8(action_items[i].character, utf8);
-                if (numbytes > 12)
-                {
-                    g_error("g_unichar_to_utf8 overflowing buffer");
+                keyman->marker = False;
+                if (g_unichar_type(action_items[i].character) == G_UNICODE_SURROGATE) {
+                    if (keyman->firstsurrogate == 0) {
+                        keyman->firstsurrogate = action_items[i].character;
+                        g_message("first surrogate %d", keyman->firstsurrogate);
+                    }
+                    else {
+                        glong items_read, items_written;
+                        gunichar2 utf16_pair[2] = {keyman->firstsurrogate, action_items[i].character};
+                        gchar *utf8_pair = g_utf16_to_utf8 (utf16_pair, 2,
+                            &items_read,
+                            &items_written,
+                            NULL);
+                        ibus_keyman_engine_commit_string(keyman, utf8_pair);
+                        g_free(utf8_pair);
+                        keyman->firstsurrogate = 0;
+                    }
                 }
-                if (numbytes)
-                {
-                    utf8[numbytes] = 0;
+                else {
+                    numbytes = g_unichar_to_utf8(action_items[i].character, utf8);
+                    if (numbytes > 12)
+                    {
+                        g_error("g_unichar_to_utf8 overflowing buffer");
+                    }
+                    if (numbytes)
```

Comment:
* More explicit to test `if (numbytes > 0)`
* Surely the `ibus_keyman_engine_commit_string` call should only be made if `numbytes > 0` as well; if the `utf8` buffer guaranteed to be zeroed beforehand? If not, then current pattern could result in an unterminated string sent to `ibus_keyman_engine_commit_string`

Reply:
initialising zeroed utf8